astropy/units/format/ogip.py

```diff
@@ -205,34 +205,31 @@ def p_unit_expression(p):
                             | UNIT OPEN_PAREN complete_expression CLOSE_PAREN power numeric_power
                             | OPEN_PAREN complete_expression CLOSE_PAREN power numeric_power
             """
-            # If we run p[1] in cls._functions, it will try and parse each
-            # item in the list into a unit, which is slow. Since we know that
-            # all the items in the list are strings, we can simply convert
-            # p[1] to a string instead.
-            p1_str = str(p[1])
-
-            if p1_str in cls._functions and p1_str != "sqrt":
```

Comment:
I think the real speed-up comes from moving this condition to inside the `if` statement. But with this, does `u.Unit("asin", format="ogip")` still raise the same error message?

I'm actually a bit surprised by this part of the parser: I think `UNIT` can only be a function here, right? So, why is that not made explicit?

Reply:
`asin` doesn't parse as a unit in current `main`, but there's no error either:
```python
>>> from astropy import units as u
>>> type(u.Unit("asin", format="ogip"))
<class 'str'>
```

As far as I can tell `UNIT` is indeed only allowed to be a function here, but if input is invalid then it could be something else.

Reply:
Hmm, I think you are right, really this can only happen with the parentheses, so it should be OK.